File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;

class Solution {

    public int[] solution(int n, String[] words) {

        Map<String, Boolean> spokenWords = new HashMap<>();

        int order = 1;
        int phase = 1;

        String prevWord = String.valueOf(words[0].charAt(0));

        for (String word : words) {

            if (spokenWords.containsKey(word) ||
                prevWord.charAt(prevWord.length() - 1) != word.charAt(0)) {

                return new int[]{order, phase};
            }

            if (order == n) {
                phase++;
            }

            spokenWords.put(word, true);
            prevWord = word;
            order = (order % n) + 1;
        }

        return new int[]{0, 0};
    }
}

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.js
function solution(n, words) {

    const spokenWords = new Map();

    let order = 1;
    let phase = 1;

    let prevWord = words[0][0];

    for (const word of words) {

        if (
            spokenWords.has(word) ||
            prevWord[prevWord.length - 1] !== word[0]
        ) {
            return [order, phase];
        }

        if (order === n) {
            phase++;
        }

        spokenWords.set(word, true);

        prevWord = word;
        order = (order % n) + 1;
    }

    return [0, 0];
}